src/advanced/service-testing.md

```diff
@@ -1,155 +1,260 @@
-    - [peers timeout](consensus/requests.md#peers-timeout)
+```rust
+let tx = MyTransaction::new(...);
+assert!(tx.verify());
+```
 
-- [Sending requests](consensus/requests.md#sending-requests) and
-  [processing of the requests](consensus/requests.md#requests-processing)
+Testkit also allows to check different orderings of transactions, including transactions
+for multiple services. This could allow to more efficiently test margin cases,
+which would be difficult to reproduce otherwise.
 
-- Acting as a [round leader](../architecture/consensus.md#strawman-version)
+```rust
+let mut testkit = TestKitBuilder::validator()
+    .with_service(MyService::new())
+    .with_service(OtherService::new())
+    .create();
+// Create transactions.
+let tx1 = MyTransaction::new(...);
+let tx2 = OtherTransaction::new(...);
+// Commit it into blockchain.
+testkit.create_block_with_transactions(txvec![tx1, tx2]);
+// Check the expected result.
+```
 
-!!! tip
-    See [source code](https://github.com/exonum/exonum/blob/master/sandbox/tests/consensus.rs)
-    for more details on how sandbox is used for testing the consensus algorithm.
+### API testing
 
-!!! note
-    The validator numbers correspond to the validator keys in the
-    [`validators` list](../architecture/configuration.md#genesis) specified in
-    the global configuration.
+The basic workflow for testing API endpoints of an Exonum service
+with the testkit is as follows:
 
-The example below shows how sandbox could be used to check `Connect` message
-exchange.
+1. Define the `MyServiceApi` trait for the `TestKitApi` structure that covers
+  the whole API of your service.
+2. Implement functions that use some transactions as test data to fill the storage.
+3. Create the tests that check all of your endpoints.
 
 ```rust
-#[test]
-fn test_sandbox_recv_and_send() {
-    let s = sandbox_with_services(vec![Box::new(ConfigUpdateService::new())]);
-    let (public, secret) = gen_keypair();
+// API trait definition.
+trait MyServiceApi {
+    fn get_public_data(&self) -> PublicDataResponse;
+    fn get_private_data(&self) -> PrivateDataResponse;
+    fn post_private_data(&self, data: &PrivateData)
+        -> PostPrivateDataResponse;
+}
 
-    // Simulate receiving "Connect" message by the node
-    s.recv(Connect::new(&public, s.a(2), s.time(), &secret));
+impl MyServiceApi for TestKitApi {
+    fn get_public_data(&self) -> PublicDataResponse {
+        self.get(
+            ApiKind::Service("my_service"),
+            "/v1/first_endpoint",
+        )
+    }
+
+    fn get_private_data(&self) -> PrivateDataResponse {
+        self.get_private(
+            ApiKind::Service("my_service"),
+            "/v1/second_endpoint",
+        )
+    }
+
+    fn post_private_data(&self, data: &PrivateData)
+        -> PostPrivateDataResponse
+    {
+        self.post(
+            ApiKind::Service("my_service"),
+            "v1/third_endpoint",
+            data,
+        )
+    }
+}
 
```

Comment:
in the testkit manual this part was preceded by a "// API test skeleton" note. Is it omitted intentionally?

Reply:
Yes, I find this comment largely unnecessary.